### PR TITLE
TEE-MIP: Safeguard Miner-generated data with Enclaves

### PR DESCRIPTION
**Pull request by mudler**

### CODE REVIEW

File: MIP/001-tee-scraper.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
+# MIP-#: Safeguard Miner-generated data with Enclaves
```
Comment (mudler) [inline MIP/001-tee-scraper.md]:
```suggestion
# MIP-1: Safeguard Miner-generated data with Enclaves
```

**Review by mcamou — COMMENTED** [review]
This is a great idea, I have a few questions/comments:

* Have we done any research on the potential impact on the network of requiring hardware equipped with a TEE? Do we know what fraction of current miners have this sort of hardware and what fraction of those who don't would be open to the expense of upgrading?
* I have only a superficial knowledge of how a TEE works, so this might be a dumb question. Have we given thought to managing encryption key distribution, rotation and revocation?
* I understand that the main purposes are to validate the integrity of the scraper code and the integrity and provenance of the data (provenance meaning that it comes from our scraper, not necessarily from its point of origin). Since we are not concerned with confidentiality, wouldn't a signature be more appropriate than full encryption?

**Comment by mudler** [comment]
> This is a great idea, I have a few questions/comments:
> 
>     * Have we done any research on the potential impact on the network of requiring hardware equipped with a TEE? Do we know what fraction of current miners have this sort of hardware and what fraction of those who don't would be open to the expense of upgrading?

A very good point.

We basically assume that no-one is equipped with this hardware locally. In the document there are listed several cloud offerings that provides SGX compatible hardware. Unfortunately finding local hardware with the Intel SGX support is hard, especially since it was moved under Intel Xeon.

One of the consequences of this is that someone deciding to join the network and partecipate to the incentives has to invest - and this also benefits the network as the quality of the data improves. Other networks had in the past similar reasoning (think about ASICs for mining, GPUs, etc): in our case however, we don't throw the computation away to calculate a proof of work, we instead use these instructions to make sure that the data transiting in the network is of good quality and trusted.


> 
>     * I have only a superficial knowledge of how a TEE works, so this might be a dumb question. Have we given thought to managing encryption key distribution, rotation and revocation?

That's a miss from my side in this document, thanks for pointing that out. Let me improve it so this is more clear!

> 
>     * I understand that the main purposes are to validate the integrity of the scraper code and the integrity and provenance of the data (provenance meaning that it comes from our scraper, not necessarily from its point of origin). Since we are not concerned with confidentiality, wouldn't a signature be more appropriate than full encryption?

Correct, however I didn't went into that direction mainly two reasons:

- Verifying signatures in TEE is provided by Remote attestation functionalities. That is actually more complex to setup as it requires to configure a quote provider: https://ego-docs.netlify.app/ego/reference/attest#set-up-the-quote-provider. :shrug: 
- In the long term, we want to seal the data generated from the worker as such it is available only to the client who originated the request, so encryption is a good fit.  We want to improve the security of the network, and similarly we will also be able to re-use same TEE bits to make sure that we write data in the Content addressable system which is only accessible by our signed binaries, to further improve the security of the network

**Comment by mcamou** [comment]
> We basically assume that no-one is equipped with this hardware locally. In the document there are listed several cloud offerings that provides SGX compatible hardware. Unfortunately finding local hardware with the Intel SGX support is hard, especially since it was moved under Intel Xeon.

That would mean fewer on-site and more cloud miners. Are we concerned about the centralization risk? Thinking about having a large percentage of miners in e.g. AWS.

> One of the consequences of this is that someone deciding to join the network and participate to the incentives has to invest - and this also benefits the network as the quality of the data improves. Other networks had in the past similar reasoning (think about ASICs for mining, GPUs, etc): in our case however, we don't throw the computation away to calculate a proof of work, we instead use these instructions to make sure that the data transiting in the network is of good quality and trusted.

This does make sense, however I would be concerned given the current and targeted size of the miner network.

> * Verifying signatures in TEE is provided by Remote attestation functionalities. That is actually more complex to setup as it requires to configure a quote provider: [ego-docs.netlify.app/ego/reference/attest#set-up-the-quote-provider](https://ego-docs.netlify.app/ego/reference/attest#set-up-the-quote-provider). 🤷

From my reading I was under the impression that remote attestation has to do with verifying the app code code. I was thinking more about signing the messages themselves to ensure their provenance and integrity, which are a separate concern (and could be done using Go libraries if the TEE does not expose the appropriate APIs).

> * In the long term, we want to seal the data generated from the worker as such it is available only to the client who originated the request, so encryption is a good fit.  We want to improve the security of the network, and similarly we will also be able to re-use same TEE bits to make sure that we write data in the Content addressable system which is only accessible by our signed binaries, to further improve the security of the network

Short term, which key do we use for encryption? Using the client's public key gives us confidentiality, while using the worker's private key allows us to validate the data provenance (given some reliable and secure way of distributing the public keys). I think it would make sense to separate out the different concerns (provenance, integrity and confidentiality) and deal with them separately, since their solutions are orthogonal to each other (strictly speaking, verifying integrity depends on verifying provenance).

Also, from my reading it would appear that each TEE generates its own Enclave Product Key, meaning that the client's scraper binary would not have access to the worker's key (which brings us again to key distribution). I might be misreading the fourth bullet in the Specification section.

If I'm going too deeply into the details for the purposes of this document, feel free to tell me to blow off :) .

**Comment by mudler** [comment]
> > We basically assume that no-one is equipped with this hardware locally. In the document there are listed several cloud offerings that provides SGX compatible hardware. Unfortunately finding local hardware with the Intel SGX support is hard, especially since it was moved under Intel Xeon.
> 
> That would mean fewer on-site and more cloud miners. Are we concerned about the centralization risk? Thinking about having a large percentage of miners in e.g. AWS.

Actually there are quite few companies providing Intel SGX (probably not an exhaustive list):

- Alibaba
- OVH
- Hetzner
- Azure
- IBM

From this side I wouldn't be really worried. If someone wants to take it further can buy an Intel Xeon processor and mine at home. If we have to compare it with mining on GPUs, it's going to be much cheaper (a Xeon CPU can also be expensive just as 1/2k)

> 
> > One of the consequences of this is that someone deciding to join the network and participate to the incentives has to invest - and this also benefits the network as the quality of the data improves. Other networks had in the past similar reasoning (think about ASICs for mining, GPUs, etc): in our case however, we don't throw the computation away to calculate a proof of work, we instead use these instructions to make sure that the data transiting in the network is of good quality and trusted.
> 
> This does make sense, however I would be concerned given the current and targeted size of the miner network.

What's the concern? I think it's totally reasonable to ask dedicated hardware to improve the quality of the data circulating over the network - given that the data is what makes the network valuable.

Let's take example from ETH(pre-pow) or Bitcoin: miners just bought dedicated GPU to mine the token, as it was impracticable to use on standard CPUs (except for Monero, right). 

I don't think that's a barrier if the value we provide it's more worth then the setup costs, as history shows.

Besides, there are other projects using TEE technologies in a similar manner, now I can just find only one reference I have bookmarked, but when I was digging there where quite few solutions out ( https://github.com/hyperledger/fabric-private-chaincode/blob/main/ecc_go/README.md )

Also see for more context: https://github.com/erayack/awesome-sgx-blockchain

> 
> > * Verifying signatures in TEE is provided by Remote attestation functionalities. That is actually more complex to setup as it requires to configure a quote provider: [ego-docs.netlify.app/ego/reference/attest#set-up-the-quote-provider](https://ego-docs.netlify.app/ego/reference/attest#set-up-the-quote-provider). 🤷
> 
> From my reading I was under the impression that remote attestation has to do with verifying the app code code. I was thinking more about signing the messages themselves to ensure their provenance and integrity, which are a separate concern (and could be done using Go libraries if the TEE does not expose the appropriate APIs).

Remote attestation verifies that who runs the code:

- Runs a signed binary
- The runtime environment has a matching measurement with pre-calculated hashes so the enclave it's not tampered with
- Ensure authenticity of the signer

The RA works by challenges with signatures, and that's the only way to leverage signatures (from my investigation with the ego SDK framework). My first attempt was actually using this methodology, but suddenly turned out to not be appropriate for the complex setup (the quote server which is needed).

It simply boils to the point that signing messages as we are used with openssl is "not implemented" in the ego SDK, and I didn't went down the rabbit hole to check if Open Enclave supports it, because that would mean more efforts for us to support messages with signatures (as we would need to surface such a feature to ego first).

Check e.g. docs at: https://pkg.go.dev/github.com/edgelesssys/ego@v1.6.0/ecrypto

> 
> > * In the long term, we want to seal the data generated from the worker as such it is available only to the client who originated the request, so encryption is a good fit.  We want to improve the security of the network, and similarly we will also be able to re-use same TEE bits to make sure that we write data in the Content addressable system which is only accessible by our signed binaries, to further improve the security of the network
> 
> Short term, which key do we use for encryption? Using the client's public key gives us confidentiality, while using the worker's private key allows us to validate the data provenance (given some reliable and secure way of distributing the public keys). I think it would make sense to separate out the different concerns (provenance, integrity and confidentiality) and deal with them separately, since their solutions are orthogonal to each other (strictly speaking, verifying integrity depends on verifying provenance).

We use product keys which are associated to the binary publisher: that means that only binaries which are signed with our key can decrypt the content which was generated into the tee-enclave.  How it works: we generate a keypair (private/public key) and we use the private key to sign the binary.

The keys that we use then is not the one of the worker neither of the client - we encrypt the data in transit with the publisher's keys (MASA keys). Down in the stack, libp2p supports e2e encryption which would provide "safety" over the wire. 

We end up to verify the integrity and the provenance to be from binaries which are signed by us. There is no confidentiality guaranteed in this approach because **anyone** could decrypt the data in transit by running our signed tee-worker and asking to decrypt the data. However,  you can't decrypt the same data by generating your own keys and re-compiling the binary. The only way to decrypt that data is using our signed software.

The client public key wouldn't give us much benefits overall, because our goal currently is just to make sure that only **binary signed by us** generates data which is accepted by the network. Any other packet circulating which can't be unsealed by the product key won't be accepted.

> 
> Also, from my reading it would appear that each TEE generates its own Enclave Product Key, meaning that the client's scraper binary would not have access to the worker's key (which brings us again to key distribution). I might be misreading the fourth bullet in the Specification section.
> 
> If I'm going too deeply into the details for the purposes of this document, feel free to tell me to blow off :) .

The product key is tied to the keypair which is used to sign the binary, different keypair, different product ID.

**Comment by mcamou** [comment]
I see. So we don't care which miner generated the data, only that it was generated by our signed binary. It might be a good idea to spell this out in the document, to prevent future confusion. When talking about security I find it useful to be very explicit regarding exactly what the threat model is and what the solution is and is not doing.

My comment regarding the client and confidentiality was regarding the future possibility of sealing the data so that it can only be decrypted by the client.

As long as we're aware of a possible centralization risk, that some current miners might be disincentivized to continue mining, and that we're raising the bar for new miners, we should be fine. I just wanted to talk through the tradeoffs and make sure that we're aware of them.

The only issue left that I see is how to distribute the Masa keys, and how to update and revoke them if/when they are compromised, especially since we'll need to somehow add the Masa product keys to all miners' TEEs.

**Review by mcamou — APPROVED** [review]

**Comment by mudler** [comment]
Thanks for the review @mcamou !

**Pull request merged**